File: dart_cli/lib/ZstBlocksFile.dart

import "dart:typed_data";

import "package:es_compression/zstd.dart";

import "RandomAccessFileWrapper.dart";

const _endian = Endian.little;

class ZstBlocksFile {
  final List<ZstBlock> blocks;

  const ZstBlocksFile(this.blocks);

  static Future<Uint8List> readBlockRowAt(RandomAccessFileWrapper file, int blockPosition, int rowIndex) async {
    await file.setPosition(blockPosition);
    return await ZstBlock.readRow(file, rowIndex);
  }

  static Stream<Uint8List> streamRows(RandomAccessFileWrapper file) async* {
    var fileSize = await file.length;
    while (file.position < fileSize) {
      yield* ZstBlock.streamRows(file);
    }
  }

  static Future<void> appendBlock(RandomAccessFileWrapper file, List<Uint8List> rows) async {
    await file.setPosition(await file.length);
    await ZstBlock(rows).write(file);
  }

  static Future<void> writeStream(RandomAccessFileWrapper file, Stream<Uint8List> rowStream, int blockSize) async {
    List<Uint8List> pendingRows = [];
    await for (var row in rowStream) {
      pendingRows.add(row);
      if (pendingRows.length >= blockSize) {
        await ZstBlock(pendingRows).write(file);
        pendingRows = [];
      }
    }
    if (pendingRows.isNotEmpty)
      await ZstBlock(pendingRows).write(file);
  }

  static Future<void> writeBlocksStream(RandomAccessFileWrapper file, Stream<List<Uint8List>> blocksStream) async {
    await for (var rows in blocksStream) {
      await ZstBlock(rows).write(file);
    }
  }
}

/// struct ZstCompressedBlock {
/// 	uint32 compressedSize;
/// 	byte compressedData[compressedSize];
/// }
/// 
/// struct ZstDecompressedBlock {
/// 	uint32 count;
/// 	ZstBlockEntry entries[count];
/// 	BlockRow rows[count];
/// }
class ZstBlock {
  final List<Uint8List> rows;

  const ZstBlock(this.rows);

  static Future<ZstBlock> fromStream(Stream<Uint8List> stream) async {
    return ZstBlock(await stream.toList());
  }

  static Stream<Uint8List> streamRows(RandomAccessFileWrapper file) async* {
    var blockSize = await file.readUint32();
    var compressedData = await file.readBytes(blockSize);
    var decompressedData = zstd.decode(compressedData);
    var bytes = ByteData(decompressedData.length);
    bytes.buffer.asUint8List().setAll(0, decompressedData);
    
    var count = bytes.getUint32(0, _endian);
    var entries = List.generate(
      count,
      (i) => ZstBlockEntryInfo.read(bytes, 4 + i * ZstBlockEntryInfo.structSize)
    );
    
    var dataStart = 4 + count * ZstBlockEntryInfo.structSize;
    for (var entry in entries) {
      yield bytes.buffer.asUint8List(dataStart + entry.offset, entry.size);
    }
  }

  static Future<Uint8List> readRow(RandomAccessFileWrapper file, int rowIndex) async {
    var blockSize = await file.readUint32();
    var compressedData = await file.readBytes(blockSize);
    var decompressedData = zstd.decode(compressedData);
    var bytes = ByteData(decompressedData.length);
    bytes.buffer.asUint8List().setAll(0, decompressedData);
    
    var count = bytes.getUint32(0, _endian);
    if (rowIndex >= count)
      throw Exception("Entry index out of range");
    var row = ZstBlockEntryInfo.read(bytes, 4 + rowIndex * ZstBlockEntryInfo.structSize);

    var dataStart = 4 + count * ZstBlockEntryInfo.structSize;
    return bytes.buffer.asUint8List(dataStart + row.offset, row.size);
  }

  Future<void> write(RandomAccessFileWrapper file) async {
    var uncompressedSize = 
      4 +
      rows.length * ZstBlockEntryInfo.structSize +
      rows.fold<int>(0, (sum, row) => sum + row.length);
    var uncompressedBytes = ByteData(uncompressedSize);
    uncompressedBytes.setUint32(0, rows.length, _endian);
    
    var dataOffset = 4 + rows.length * ZstBlockEntryInfo.structSize;
    var currentDataLocalOffset = 0;
    for (int i = 0; i < rows.length; i++) {
      var row = rows[i];
      var entryInfo = ZstBlockEntryInfo(currentDataLocalOffset, row.length);
      entryInfo.write(uncompressedBytes, 4 + i * ZstBlockEntryInfo.structSize);
      uncompressedBytes.buffer.asUint8List(dataOffset + currentDataLocalOffset, row.length).setAll(0, row);
      currentDataLocalOffset += row.length;
    }
    var uncompressedData = uncompressedBytes.buffer.asUint8List();
    var compressedData = zstd.encode(uncompressedData);
    var compressedSize = compressedData.length;
    var blockBytes = ByteData(4 + compressedSize);
    blockBytes.setUint32(0, compressedSize, _endian);
    blockBytes.buffer.asUint8List(4).setAll(0, compressedData);
    await file.write(blockBytes);
  }
}

class ZstBlockEntryInfo {
  static const int structSize = 8;
  final int offset;
  final int size;

  const ZstBlockEntryInfo(this.offset, this.size);

  static ZstBlockEntryInfo read(ByteData bytes, int position) {
    var offset = bytes.getUint32(position + 0, _endian);
    var size = bytes.getUint32(position + 4, _endian);
    return ZstBlockEntryInfo(offset, size);
  }

  void write(ByteData bytes, int position) {
    bytes.setUint32(position + 0, offset, _endian);
    bytes.setUint32(position + 4, size, _endian);
  }
}

File: python_cli/ZstBlocksFile.py
from __future__ import annotations
import os
from typing import BinaryIO, Iterable
from zstandard import ZstdDecompressor, ZstdCompressor

_endian = "little"

class ZstBlocksFile:
	blocks: list[ZstBlock]

	def __init__(self, blocks: list[ZstBlock]):
		self.blocks = blocks

	@staticmethod
	def readBlockRowAt(file: BinaryIO, blockPosition: int, rowIndex: int) -> bytes:
		file.seek(blockPosition)
		return ZstBlock.readRow(file, rowIndex)
	
	@staticmethod
	def streamRows(file: BinaryIO) -> bytes:
		fileSize = os.path.getsize(file.name)
		while file.tell() < fileSize:
			yield from ZstBlock.streamRows(file)
	
	@staticmethod
	def appendBlock(file: BinaryIO, rows: list[bytes]) -> None:
		file.seek(file.tell())
		ZstBlock(rows).write(file)
	
	@staticmethod
	def writeStream(file: BinaryIO, rowStream: Iterable[bytes], blockSize: int) -> None:
		pendingRows = []
		for row in rowStream:
			pendingRows.append(row)
			if len(pendingRows) >= blockSize:
				ZstBlock(pendingRows).write(file)
				pendingRows = []
		if len(pendingRows) > 0:
			ZstBlock(pendingRows).write(file)

	@staticmethod
	def writeBlocksStream(file: BinaryIO, blocksStream: Iterable[list[bytes]]) -> None:
		for rows in blocksStream:
			ZstBlock(rows).write(file)
	

class ZstBlock:
	rows: list[bytes]

	def __init__(self, rows: list[bytes]):
		self.rows = rows

	@staticmethod
	def fromStream(stream: Iterable[bytes]) -> ZstBlock:
		return ZstBlock(list(stream))

	@classmethod
	def streamRows(cls, file: BinaryIO) -> bytes:
		blockSize = int.from_bytes(file.read(4), _endian)
		compressedData = file.read(blockSize)
		stream = ZstdDecompressor().stream_reader(compressedData)
		decompressedData = stream.read()
		stream.close()
		bytes = bytearray(decompressedData)
		
		count = int.from_bytes(bytes[0:4], _endian)
		entries = [ZstBlockEntryInfo.read(bytes, 4 + i * ZstBlockEntryInfo.structSize) for i in range(count)]
		
		dataStart = 4 + count * ZstBlockEntryInfo.structSize
		for entry in entries:
			yield bytes[dataStart + entry.offset : dataStart + entry.offset + entry.size]
	
	@classmethod
	def readRow(cls, file: BinaryIO, rowIndex: int) -> bytes:
		blockSize = int.from_bytes(file.read(4), _endian)
		compressedData = file.read(blockSize)
		stream = ZstdDecompressor().stream_reader(compressedData)
		decompressedData = stream.read()
		stream.close()
		bytes = bytearray(decompressedData)
		
		count = int.from_bytes(bytes[0:4], _endian)
		if rowIndex >= count:
			raise Exception("Entry index out of range")
		row = ZstBlockEntryInfo.read(bytes, 4 + rowIndex * ZstBlockEntryInfo.structSize)

		dataStart = 4 + count * ZstBlockEntryInfo.structSize
		return bytes[dataStart + row.offset : dataStart + row.offset + row.size]

	def write(self, file: BinaryIO) -> None:
		uncompressedSize = \
			4 + \
			len(self.rows) * ZstBlockEntryInfo.structSize + \
			sum(len(row) for row in self.rows)
		uncompressedBytes = bytearray(uncompressedSize)
		uncompressedBytes[0:4] = len(self.rows).to_bytes(4, _endian)
		
		dataOffset = 4 + len(self.rows) * ZstBlockEntryInfo.structSize
		currentDataLocalOffset = 0
		for i in range(len(self.rows)):
			row = self.rows[i]
			entryInfo = ZstBlockEntryInfo(currentDataLocalOffset, len(row))
			entryInfo.write(uncompressedBytes, 4 + i * ZstBlockEntryInfo.structSize)
			uncompressedBytes[dataOffset + currentDataLocalOffset : dataOffset + currentDataLocalOffset + len(row)] = row
			currentDataLocalOffset += len(row)
		uncompressedData = bytes(uncompressedBytes)
		compressedData = ZstdCompressor().compress(uncompressedData)
		compressedSize = len(compressedData)
		blockBytes = bytearray(4 + compressedSize)
		blockBytes[0:4] = compressedSize.to_bytes(4, _endian)
		blockBytes[4:4+compressedSize] = compressedData
		file.write(blockBytes)

class ZstBlockEntryInfo:
	structSize = 8
	offset: int
	size: int

	def __init__(self, offset: int, size: int):
		self.offset = offset
		self.size = size

	@staticmethod
	def read(bytes: bytearray, position: int) -> ZstBlockEntryInfo:
		offset = int.from_bytes(bytes[position + 0 : position + 4], _endian)
		size = int.from_bytes(bytes[position + 4 : position + 8], _endian)
		return ZstBlockEntryInfo(offset, size)

	def write(self, bytes: bytearray, position: int) -> None:
		bytes[position + 0 : position + 4] = self.offset.to_bytes(4, _endian)
		bytes[position + 4 : position + 8] = self.size.to_bytes(4, _endian)
